### PR TITLE
INT-6697 - Initial implementation

### PR DESCRIPTION
**Pull request by adam-in-ict**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Summary

Initial BigID implementation

### Entities

The following entities are created:

| Resources   | Entity `_type`     | Entity `_class`  |
| ----------- | ------------------ | ---------------- |
| Account     | `bigid_account`    | `Account`        |
| Data Source | `bigid_datasource` | `DataCollection` |
| PII Object  | `bigid_pii_object` | `Record`         |
| User        | `bigid_user`       | `User`           |

### Relationships

The following relationships are created:

| Source Entity `_type` | Relationship `_class` | Target Entity `_type` |
| --------------------- | --------------------- | --------------------- |
| `bigid_account`       | **HAS**               | `bigid_user`          |
| `bigid_account`       | **SCANS**             | `bigid_datasource`    |
| `bigid_datasource`    | **HAS**               | `bigid_pii_object`    |

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### General Development Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [ ] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [ ] My changes properly paginate the target service provider's API
- [ ] My changes properly handle rate limiting of the target service provider's
      API
- [ ] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [ ] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [ ] I have updated the `CHANGELOG.md` file to describe my changes
- [ ] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,2 +1,3 @@
-CLIENT_ID=
-CLIENT_SECRET=
\ No newline at end of file
+BASE_URL=
+USERNAME=
```
Comment (zemberdotnet) [inline .env.example:2]: We shouldn't use `USERNAME` anymore: 
https://github.com/JupiterOne/sdk/pull/817

File: .github/workflows/questions.yml [hunk 2]
```diff
@@ -24,9 +24,10 @@ jobs:
         with:
           node-version: 14.x
           cache: yarn
+          cache-dependency-path: '**/yarn.lock'
```
Comment (zemberdotnet) [inline .github/workflows/questions.yml:27]: Should we contribute this to the template?

File: src/config.ts [hunk 3]
```diff
@@ -4,59 +4,49 @@ import {
...
- * managed environment. For example, in JupiterOne, users configure
- * `instance.config` in a UI.
- */
 export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
-  clientId: {
+  baseUrl: {
     type: 'string',
   },
-  clientSecret: {
+  username: {
```
Comment (zemberdotnet) [inline src/config.ts:13]: Same as other comment above.

File: src/steps/dataSource/converter.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        type: source.type,
+        scanStrategy: source.scanner_strategy,
+        // These two are string instead of epoch like last_scan_at.
+        createdOn: parseTimePropertyValue(source.created_at),
+        updatedOn: parseTimePropertyValue(source.updated_at),
+        active: source.enabled == 'yes',
+        // The owners array contains complex objects we can't directly map
+        // to a property, so reduce it down to an array of email addresses.
+        owners: source.owners_v2?.map((item) => item.email),
+        // the API doesn't provide a classification, but this is a required field
```
Comment (zemberdotnet) [inline src/steps/dataSource/converter.ts:43]: https://github.com/JupiterOne/data-model/blob/f028a3cd7c1c734bb9865ee2784c18589a4daf9e/src/schemas/Entity.json#L150

I think this strategy is good but we should change `owners` to `owner`.

File: src/steps/findings/converter.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { FindingRow } from '../../types';
+
+function createFindingKey(source: string, name: string) {
+  return `${Entities.FINDING._type}:${source}${name}`;
+}
+
+export function createFindingEntity(finding: FindingRow): Entity {
+  return createIntegrationEntity({
+    entityData: {
+      source: [],
```
Comment (zemberdotnet) [inline src/steps/findings/converter.ts:20]: Also intentionally leaving out rawData here?

File: src/steps/findings/index.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      createFindingEntity(finding),
+    );
+    const sourceEntity = await jobState.findEntity(
+      createDataSourceKey(finding['Data Source']),
+    );
+
+    if (findingEntity && sourceEntity) {
+      await jobState.addRelationship(
+        createSourceFindingRelationship(sourceEntity, findingEntity),
+      );
```
Comment (zemberdotnet) [inline src/steps/findings/index.ts:33]: Should we add a log when we fail to findEntity or create the relationship? Also slight simplification:
```suggestion
    const sourceEntity = await jobState.findEntity(
      createDataSourceKey(finding['Data Source']),
    );

    if (sourceEntity) {
      await jobState.addRelationship(
        createSourceFindingRelationship(sourceEntity, findingEntity),
      );
```

File: .github/workflows/questions.yml (hunk 2 above)
Comment (adam-in-ict) [reply .github/workflows/questions.yml:27]: I still can't get the test to pass, so didn't contribute it back in yet.

File: src/steps/findings/converter.ts (hunk 5 above)
Comment (adam-in-ict) [reply src/steps/findings/converter.ts:20]: Yes.  Their documentation made me unsure if we could guarantee there'd never be sensitive information passed back.

File: .env.example (hunk 1 above)
Comment (adam-in-ict) [reply .env.example]: Thanks for the catch!  Updated in latest commits.

**Review by zemberdotnet — APPROVED** [review]

**Pull request merged**